test/test_shared.cpp

```diff
@@ -2080,14 +2108,104 @@ TEST_IF(Shared_AsyncMultiprocess, allow_async)
+#ifdef _WIN32
+
+#if 1
+
+TEST(Shared_WaitForChangeAfterOwnCommit)
+{
+    SHARED_GROUP_TEST_PATH(path);
+
+    SharedGroup* sg = new SharedGroup(path);
+    sg->begin_write();
+    sg->commit();
+    bool b = sg->wait_for_change();
+}
+
+
+TEST(Shared_InterprocessWaitForChange)
+{
+    // We can't use SHARED_GROUP_TEST_PATH() because it will attempt to clean up the .realm file at the end,
+    // and hence throw if the other processstill has the .realm file open
+    std::string path = get_test_path("Shared_InterprocessWaitForChange", ".realm");
+
+    // This works differently from POSIX: Here, the child process begins execution from the start of this unit
+    // test and not from the place of fork().
+    DWORD pid = winfork("Shared_InterprocessWaitForChange");
+
+    if (pid == -1) {
+        CHECK(false);
+        return;
+    }
 
+    std::unique_ptr<SharedGroup> sg(new SharedGroup(path));
+
+    // An old .realm file with random contents can exist (such as a leftover from earlier crash) with random
+    // data, so we always initialize the database
+    {
+        Group& g = sg->begin_write();
+        if (g.size() == 1) {
+            g.remove_table("data");
+            TableRef table = g.add_table("data");
+            table->add_column(type_Int, "ints");
+            table->add_empty_row();
+            table->set_int(0, 0, 0);
+        }
+        sg->commit();
+        sg->wait_for_change();
+    }
+
+    bool first = false;
+    fastrand(time(0), true);
+
+    // By turn, incremenet the counter and wait for the other to increment it too
+    for (int i = 0; i < 10; i++)
+    {
+        Group& g = sg->begin_write();
+        if (g.size() == 1) {
+            TableRef table = g.get_table("data");
+            int64_t v = table->get_int(0, 0);
+
+            if (i == 0 && v == 0)
+                first = true;
+
+            // Note: If this fails in child process (pid != 0) it might go undetected. This is not
+            // critical since it will most likely result in a failure in the parent process also.
+            CHECK_EQUAL(v - (first ? 0 : 1), 2 * i);
+            table->set_int(0, 0, v + 1);
+        }
+
+        // millisleep(0) might yield time slice on certain OS'es, so we use fastrand() to get cases 
+        // of 0 delay, because non-yieldig is also an important test case.
+        if(fastrand(1))
+            millisleep((time(0) % 10) * 10);
+
+        sg->commit();
+
+        if (fastrand(1))
+            millisleep((time(0) % 10) * 10);
+
+        sg->wait_for_change();
+
+        if (fastrand(1))
+            millisleep((time(0) % 10) * 10);
+    }
+
+    // Wake up other process so it will exit too
+    sg->begin_write();
+    sg->commit();
+}
+
+#endif
+
+#endif
+
+// FIXME: This test does not work with valgrind, but we still run it on Windows
 // This test will hang infinitely instead of failing!!!
+#ifdef _WIN32
 TEST(Shared_WaitForChange)
 {
-    const int num_threads = 3;
+    const int num_threads = 1;
```

Comment:
Oops, old leftover from messing around. Btw. I decided to enable the test on *nix not despite of valgrind. The test is important, especially after Finn's update to that method